## ZAWANSOWANE SYSTEMY SZTUCZNEJ INTELIGENCJI

Wykonawcy: Kacper Kalinowski, Aleksander Kalinowski, Sylwia Mościcka

Zbiór danych

In [1]:
import pandas as pd
%matplotlib inline

In [2]:
df = pd.read_csv("Houses.csv", encoding="windows-1250")
df

Unnamed: 0                                            address  \
0               0           Podgórze Zabłocie Stanisława Klimeckiego   
1               1                          Praga-Południe Grochowska   
2               2                            Krowodrza Czarnowiejska   
3               3                                           Grunwald   
4               4  Ochota Gotowy budynek. Stan deweloperski. Osta...   
...           ...                                                ...   
23759       23759                            Stare Miasto Naramowice   
23760       23760                                             Włochy   
23761       23761                    Nowe Miasto Malta ul. Katowicka   
23762       23762                 Podgórze Duchackie Walerego Sławka   
23763       23763                                            Ursynów   

           city  floor       id   latitude  longitude      price  rooms  \
0        Kraków    2.0  23918.0  50.049224  19.970379   749000.0    3.0   
1      Warszawa    3.0  17828.0  52.249775  21.106886   240548.0    1.0   
2        Kraków    2.0  22784.0  50.066964  19.920025   427000.0    2.0   
3        Poznań    2.0   4315.0  52.404212  16.882542  1290000.0    5.0   
4      Warszawa    1.0  11770.0  52.212225  20.972630   996000.0    5.0   
...         ...    ...      ...        ...        ...        ...    ...   
23759    Poznań    0.0   3976.0  52.449649  16.949408   543000.0    4.0   
23760  Warszawa    4.0  10206.0  52.186109  20.948438   910000.0    3.0   
23761    Poznań    0.0   4952.0  52.397345  16.961939   430695.0    3.0   
23762    Kraków    6.0  24148.0  50.024231  19.959569   359000.0    2.0   
23763  Warszawa    2.0  10889.0  52.141039  21.032321   604800.0    3.0   

           sq    year  
0       74.05  2021.0  
1       24.38  2021.0  
2       37.00  1970.0  
3      166.00  1935.0  
4      105.00  2020.0  
...       ...     ...  
23759   77.00  2020.0  
23760   71.00  2017.0  
23761   50.67  2022.0  
23762   38.86  2021.0  
23763   63.00  1978.0  

[23764 rows x 11 columns]

1. Usuwanie zbędnych atrybutów:
Usunięcie atrybutu "Unnamed: 0" tzw. indexu z danych, ponieważ nie wnosi on istotnych informacji i nie ma wpływu na cenę nieruchomości.

In [3]:
df = df.drop(columns=['Unnamed: 0'])

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23764 entries, 0 to 23763
Data columns (total 10 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   address    23764 non-null  object 
 1   city       23764 non-null  object 
 2   floor      23764 non-null  float64
 3   id         23764 non-null  float64
 4   latitude   23764 non-null  float64
 5   longitude  23764 non-null  float64
 6   price      23764 non-null  float64
 7   rooms      23764 non-null  float64
 8   sq         23764 non-null  float64
 9   year       23764 non-null  float64
dtypes: float64(8), object(2)
memory usage: 1.8+ MB


Usunięcie niepotrzebnych atrybutów pomaga zmniejszyć liczbę wymiarów danych, co może poprawić wydajność modelu i zmniejszyć ryzyko przeuczenia (overfitting).

2. Ekstrakcja informacji z daty:
Z atrybutu "year" można wyodrębnić informacje o dniu, miesiącu i roku, co może być przydatne w analizie zmian cen w zależności od pory roku.

Ekstrakcja informacji z daty pozwala uwzględnić sezonowe zmiany cen nieruchomości. Niestety w naszych danych nie mamy informacji o miesiacach, przez co nie jesteśmy w stanie sprawdzić, w jakich miesiacach najchętniej kupowano mieszkania.

Pozostałe dane też stworzone są w analogiczny sposób, więc nic nie musimy poddawać ekstracji.

3. Przystosowanie typu danych i kodowanie kategorii.

In [5]:
df = pd.get_dummies(df, columns=['city', 'address'])

In [14]:
df

floor   latitude  longitude      price  rooms      sq  year  \
0        2.0  50.049224  19.970379   749000.0    3.0   74.05  1970   
1        3.0  52.249775  21.106886   240548.0    1.0   24.38  1970   
2        2.0  50.066964  19.920025   427000.0    2.0   37.00  1970   
3        2.0  52.404212  16.882542  1290000.0    5.0  166.00  1970   
4        1.0  52.212225  20.972630   996000.0    5.0  105.00  1970   
...      ...        ...        ...        ...    ...     ...   ...   
23759    0.0  52.449649  16.949408   543000.0    4.0   77.00  1970   
23760    4.0  52.186109  20.948438   910000.0    3.0   71.00  1970   
23761    0.0  52.397345  16.961939   430695.0    3.0   50.67  1970   
23762    6.0  50.024231  19.959569   359000.0    2.0   38.86  1970   
23763    2.0  52.141039  21.032321   604800.0    3.0   63.00  1970   

       city_Kraków  city_Poznań  city_Warszawa  ...  \
0                1            0              0  ...   
1                0            0              1  ...   
2                1            0              0  ...   
3                0            1              0  ...   
4                0            0              1  ...   
...            ...          ...            ...  ...   
23759            0            1              0  ...   
23760            0            0              1  ...   
23761            0            1              0  ...   
23762            1            0              0  ...   
23763            0            0              1  ...   

       address_Żoliborz ul. Stanisława Tołwińskiego  \
0                                                 0   
1                                                 0   
2                                                 0   
3                                                 0   
4                                                 0   
...                                             ...   
23759                                             0   
23760                                             0   
23761                                             0   
23762                                             0   
23763                                             0   

       address_Żoliborz ul. Sybilli  address_Żoliborz ul. Szamocka  \
0                                 0                              0   
1                                 0                              0   
2                                 0                              0   
3                                 0                              0   
4                                 0                              0   
...                             ...                            ...   
23759                             0                              0   
23760                             0                              0   
23761                             0                              0   
23762                             0                              0   
23763                             0                              0   

       address_Żoliborz ul. Władysława Broniewskiego  \
0                                                  0   
1                                                  0   
2                                                  0   
3                                                  0   
4                                                  0   
...                                              ...   
23759                                              0   
23760                                              0   
23761                                              0   
23762                                              0   
23763                                              0   

       address_Żoliborz ul. Włościańska 15  \
0                                        0   
1                                        0   
2                                        0   
3                                        0   
4                                        0   
...                                    ...   
23759   

Model maszynowy wymaga atrybutów numerycznych, więc konieczne jest przekształcenie kategorii na wartości liczbowe.

4. Normalizacja wartości atrybutów.

In [6]:
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()
df[['floor','latitude', 'longitude', 'price', 'rooms', 'sq', 'year']] = scaler.fit_transform(df[['floor', 'latitude', 'longitude', 'price', 'rooms', 'sq', 'year']])


In [7]:
df

floor       id  latitude  longitude     price     rooms        sq  \
0        0.2  23918.0  0.025803   0.603707  0.049617  0.222222  0.000065   
1        0.3  17828.0  0.514577   0.647227  0.015708  0.000000  0.000015   
2        0.2  22784.0  0.029743   0.601779  0.028143  0.111111  0.000028   
3        0.2   4315.0  0.548880   0.485466  0.085695  0.444444  0.000156   
4        0.1  11770.0  0.506237   0.642086  0.066089  0.444444  0.000096   
...      ...      ...       ...        ...       ...       ...       ...   
23759    0.0   3976.0  0.558972   0.488026  0.035879  0.333333  0.000068   
23760    0.4  10206.0  0.500436   0.641160  0.060353  0.222222  0.000062   
23761    0.0   4952.0  0.547354   0.488506  0.028389  0.222222  0.000042   
23762    0.6  24148.0  0.020251   0.603294  0.023608  0.111111  0.000030   
23763    0.2  10889.0  0.490425   0.644372  0.040000  0.222222  0.000054   

           year  city_Kraków  city_Poznań  ...  \
0      0.670447            1            0  ...   
1      0.670447            0            0  ...   
2      0.652921            1            0  ...   
3      0.640893            0            1  ...   
4      0.670103            0            0  ...   
...         ...          ...          ...  ...   
23759  0.670103            0            1  ...   
23760  0.669072            0            0  ...   
23761  0.670790            0            1  ...   
23762  0.670447            1            0  ...   
23763  0.655670            0            0  ...   

       address_Żoliborz ul. Stanisława Tołwińskiego  \
0                                                 0   
1                                                 0   
2                                                 0   
3                                                 0   
4                                                 0   
...                                             ...   
23759                                             0   
23760                                             0   
23761                                             0   
23762                                             0   
23763                                             0   

       address_Żoliborz ul. Sybilli  address_Żoliborz ul. Szamocka  \
0                                 0                              0   
1                                 0                              0   
2                                 0                              0   
3                                 0                              0   
4                                 0                              0   
...                             ...                            ...   
23759                             0                              0   
23760                             0                              0   
23761                             0                              0   
23762                             0                              0   
23763                             0                              0   

       address_Żoliborz ul. Władysława Broniewskiego  \
0                                                  0   
1                                                  0   
2                                                  0   
3                                                  0   
4                                                  0   
...                                              ...   
23759                                              0   
23760                                              0   
23761                                              0   
23762                                              0   
23763                                              0   

       address_Żoliborz ul. Włościańska 15  \
0                                        0   
1                                        0   
2                                        0   
3                                        0   
4                                        0   
...                                    ...

Normalizacja pomaga uniknąć problemów związanych z różnicami w skalach atrybutów.

5. Uzupełnienie danych brakujących.

W naszym przypadku nie mieliśmy brakujących danych, przez co nie musieliśmy dodatkowo ich uzupełniać.

6. Czyszczenie danych.
Usunięcie rekordów zawierających błędne dane lub zbyt wiele brakujących wartości. Warto również pozbyć się duplikatów, aby zachować czystość danych.

In [8]:
df = df.dropna()  # Usuń wiersze z brakującymi danymi
df = df.drop_duplicates()  # Usuń duplikaty

In [9]:
df

floor       id  latitude  longitude     price     rooms        sq  \
0        0.2  23918.0  0.025803   0.603707  0.049617  0.222222  0.000065   
1        0.3  17828.0  0.514577   0.647227  0.015708  0.000000  0.000015   
2        0.2  22784.0  0.029743   0.601779  0.028143  0.111111  0.000028   
3        0.2   4315.0  0.548880   0.485466  0.085695  0.444444  0.000156   
4        0.1  11770.0  0.506237   0.642086  0.066089  0.444444  0.000096   
...      ...      ...       ...        ...       ...       ...       ...   
23759    0.0   3976.0  0.558972   0.488026  0.035879  0.333333  0.000068   
23760    0.4  10206.0  0.500436   0.641160  0.060353  0.222222  0.000062   
23761    0.0   4952.0  0.547354   0.488506  0.028389  0.222222  0.000042   
23762    0.6  24148.0  0.020251   0.603294  0.023608  0.111111  0.000030   
23763    0.2  10889.0  0.490425   0.644372  0.040000  0.222222  0.000054   

           year  city_Kraków  city_Poznań  ...  \
0      0.670447            1            0  ...   
1      0.670447            0            0  ...   
2      0.652921            1            0  ...   
3      0.640893            0            1  ...   
4      0.670103            0            0  ...   
...         ...          ...          ...  ...   
23759  0.670103            0            1  ...   
23760  0.669072            0            0  ...   
23761  0.670790            0            1  ...   
23762  0.670447            1            0  ...   
23763  0.655670            0            0  ...   

       address_Żoliborz ul. Stanisława Tołwińskiego  \
0                                                 0   
1                                                 0   
2                                                 0   
3                                                 0   
4                                                 0   
...                                             ...   
23759                                             0   
23760                                             0   
23761                                             0   
23762                                             0   
23763                                             0   

       address_Żoliborz ul. Sybilli  address_Żoliborz ul. Szamocka  \
0                                 0                              0   
1                                 0                              0   
2                                 0                              0   
3                                 0                              0   
4                                 0                              0   
...                             ...                            ...   
23759                             0                              0   
23760                             0                              0   
23761                             0                              0   
23762                             0                              0   
23763                             0                              0   

       address_Żoliborz ul. Władysława Broniewskiego  \
0                                                  0   
1                                                  0   
2                                                  0   
3                                                  0   
4                                                  0   
...                                              ...   
23759                                              0   
23760                                              0   
23761                                              0   
23762                                              0   
23763                                              0   

       address_Żoliborz ul. Włościańska 15  \
0                                        0   
1                                        0   
2                                        0   
3                                        0   
4                                        0   
...                                    ...

Czyste dane zapewniają dokładniejsze i bardziej wiarygodne wyniki modelu.

7. Podjęcie decyzji, czy zachować dane odstające (ang. outliers).

In [10]:
from sklearn.ensemble import IsolationForest

X = df[['price']]

# Inicjalizacja modelu Isolation Forest
model = IsolationForest(contamination=0.05, random_state=42)

# Dopasowanie modelu
model.fit(X)

# Oznacz anomalia w danych
df['anomaly'] = model.predict(X)
df['anomaly']

C:\Users\user\anaconda3\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but IsolationForest was fitted with feature names
  warnings.warn(


0        1
1        1
2        1
3        1
4        1
        ..
23759    1
23760    1
23761    1
23762    1
23763    1
Name: anomaly, Length: 23764, dtype: int32

Wybieramy atrybuty, na których chcemy wykrywać anomalie (w przykładzie używamy tylko 'price').
Inicjalizujemy model Isolation Forest, ustawiając parametr contamination na poziom odsetka danych uważanych za anomalie (w tym przykładzie 5%).
Dopasowujemy model do danych za pomocą fit.
Używamy modelu do przewidywania anomalii na danych i oznaczamy wyniki jako nowy atrybut 'anomaly' w naszym DataFrame.

W naszej analizie wyszło dużo anomalii, jednak ich nie usuwamy. Cena może mieć w naszym przypadku ogromne znaczenie.


Anomalie w cenach mogą wskazywać na nieruchomości, które są szczególnie atrakcyjne z ceną poniżej przeciętnej lub nieruchomości, które wymagają uwagi, ponieważ ich cena jest znacząco wyższa. To może pomóc inwestorom lub kupującym nieruchomości w identyfikacji potencjalnie atrakcyjnych ofert lub problematycznych transakcji. 

Z drugiej strony dla instytucji finansowych, które udzielają kredytów hipotecznych, identyfikacja cen nieruchomości uważanych za anomalie może pomóc w ocenie ryzyka kredytowego. Jeśli cena nieruchomości jest znacznie niższa od przeciętnej, może to oznaczać ryzyko utraty wartości nieruchomości w przypadku niewypłacalności kredytobiorcy.

In [11]:
from sklearn.ensemble import IsolationForest

X = df[['year']]

# Inicjalizacja modelu Isolation Forest
model = IsolationForest(contamination=0.05, random_state=42)

# Dopasowanie modelu
model.fit(X)

# Oznacz anomalia w danych
df['anomaly'] = model.predict(X)
anomaly_counts = df['anomaly'].value_counts()

# Wyświetl wynik
print(anomaly_counts)

C:\Users\user\anaconda3\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but IsolationForest was fitted with feature names
  warnings.warn(


 1    22652
-1     1112
Name: anomaly, dtype: int64


W roku wyniki są bardzo rozbieżne, więc również nie bedziemy usuwać danych.

In [12]:
from sklearn.ensemble import IsolationForest

X = df[['sq']]

# Inicjalizacja modelu Isolation Forest
model = IsolationForest(contamination=0.05, random_state=42)

# Dopasowanie modelu
model.fit(X)

# Oznacz anomalia w danych
df['anomaly'] = model.predict(X)
anomaly_counts = df['anomaly'].value_counts()

# Wyświetl wynik
print(anomaly_counts)

C:\Users\user\anaconda3\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but IsolationForest was fitted with feature names
  warnings.warn(


 1    22581
-1     1183
Name: anomaly, dtype: int64


Wyniki identyfikacji anomalii w atrybucie "sq" (powierzchnia nieruchomości) wskazują, że większość obserwacji (20,922) została uznana za normalne, a 1,094 obserwacje zostały uznane za anomalię.

W naszej analzie jest dużo anomali. Podjęliśmy decyzję, aby wszystkie zostawić.Zachowanie danych odstających (anomalii) w analizie danych może być uzasadnione i ważne z kilku powodów:

    Pełniejsza reprezentacja rzeczywistości: Anomalie mogą być rzeczywistymi obserwacjami, które odzwierciedlają różnorodność i nietypowość na rynku nieruchomości. Włączenie ich w analizę może pomóc w lepszym zrozumieniu pełnego zakresu danych.

    Unikalne czynniki wpływające na cenę: Nieruchomości o nietypowych rozmiarach mogą mieć unikalne cechy lub czynniki wpływające na cenę, które są ważne w analizie rynku nieruchomości. Ich uwzględnienie może dostarczyć cennych informacji.

    Identyfikacja lukratywnych okazji lub problematycznych nieruchomości: Anomalie mogą wskazywać na nieruchomości, które są szczególnie atrakcyjne (jeśli są tańsze) lub problematyczne (jeśli są droższe). To może być istotne dla inwestorów lub kupujących nieruchomości.

    Wartość predykcyjna: Anomalie mogą zawierać istotne informacje dla predykcji cen nieruchomości. Jeśli te nietypowe przypadki są częścią rzeczywistego rynku nieruchomości, to uwzględnienie ich w modelach predykcyjnych może poprawić ich skuteczność.

    Rozpoznawanie trendów rynkowych: Anomalie mogą być wskaźnikiem zmian na rynku nieruchomości. Nagłe zmiany w cenach lub rozmiarach nieruchomości mogą wskazywać na istotne trendy rynkowe.

8. Zastosowanie metod oversampling/undersampling w celu zniwelowania problemu niezbalansowania danych.

In [17]:
from sklearn.model_selection import train_test_split
X = df.drop(columns=['price'])
y = df['price']

# Podziel dane na zbiór treningowy i testowy
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Ręczny oversampling klasy mniejszościowej
class_0 = df[df['price'] == 0]
class_1 = df[df['price'] == 1]

# Określ ile próbek ma zostać dodanych w oversamplingu
n_samples_to_add = len(class_0) - len(class_1)

# Losowo wybierz próbki z klasy mniejszościowej i dodaj je wielokrotnie
oversampled_class_1 = class_1.sample(n=n_samples_to_add, replace=True, random_state=42)

# Połącz zrównoważone klasy
balanced_df_oversampled = pd.concat([class_0, oversampled_class_1])

# Ręczny undersampling klasy większościowej
n_samples_to_remove = len(class_0) - len(class_1)

# Usuń nadmiarowe próbki z klasy większościowej
undersampled_class_0 = class_0.sample(n=len(class_1), random_state=42)

# Połącz zrównoważone klasy
balanced_df_undersampled = pd.concat([undersampled_class_0, class_1])

ValueError: a must be greater than 0 unless no samples are taken

W błędzie występuje problem z próbą oversamplingu, ponieważ próbki próbowane są w ilości mniejszej lub równej zeru. To oznacza, że liczba próbek klasy mniejszościowej jest już większa lub równa liczbie próbek klasy większościowej, co sprawia, że oversampling nie jest potrzebny.

Jeśli klasy są już zrównoważone (lub klasa mniejszościowa jest większa), nie ma potrzeby stosowania oversamplingu. \

  Niezbalansowane dane: Dane mogą być niezbalansowane, co oznacza, że jedna z klas jest znacznie liczniejsza od drugiej. W Twoim przypadku, mogą istnieć nieruchomości o różnych cenach, ale pewne kategorie cenowe mogą być mniej liczne. Oversampling pomaga w zrównoważeniu klas, co jest istotne, aby model mógł nauczyć się równie dobrze klasy mniejszościowe.

    Uniknięcie przeuczenia: Przeuczenie (overfitting) to sytuacja, w której model jest zbyt dopasowany do danych treningowych i nie ogólny na nowe dane. Walidacja krzyżowa pomaga w ocenie ogólnej wydajności modelu na różnych zestawach danych treningowych i testowych, co pomaga uniknąć przeuczenia.

    Poprawa oceny modelu: Walidacja krzyżowa dostarcza średni wynik wydajności modelu na różnych zestawach danych, co pozwala na bardziej rzetelną ocenę modelu. Oversampling pomaga w lepszym wykorzystaniu danych mniejszościowych, co może poprawić wyniki modelu.

    Równowaga między precyzją a pełnością: Oversampling pomaga zrównoważyć między precyzją (dokładnością) a pełnością (pokryciem) w kontekście klasy mniejszościowej. Niezbalansowane dane mogą prowadzić do modeli, które są zbyt precyzyjne, ale nie pokrywają wszystkich przypadków.

9. Implementacja walidacji krzyżowej (uwaga: kroki opisane wyżej lub inne konieczne do przeprowadzenia w ramach przygotowania wstępnego danych należy wykonywać tylko na danych treningowych, aby uniknąć tzw. wycieku danych (ang. data leakage))

In [19]:
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
X = df.drop(columns=['price'])
y = df['price']

# Inicjalizacja modelu regresji liniowej
model = LinearRegression()

# Przeprowadzenie 5-krotnej walidacji krzyżowej
scores = cross_val_score(model, X, y, cv=5, scoring='r2')

# Wyświetlenie wyników walidacji krzyżowej
print("Wyniki walidacji krzyżowej (R^2):", scores)
print("Średni wynik walidacji krzyżowej (R^2):", scores.mean())

Wyniki walidacji krzyżowej (R^2): [-1.10132456e+13 -6.73892362e+01  4.97220934e-01 -2.92894489e+13
 -2.25321442e+10]
Średni wynik walidacji krzyżowej (R^2): -8065045337269.175


  Wyniki walidacji krzyżowej (R^2) są miarą wydajności modelu regresji liniowej na różnych podzbiorach danych. R^2 (R-squared) jest miarą, która wskazuje, jak dobrze model pasuje do danych. Wartość R^2 mieści się w zakresie od 0 do 1, gdzie:

    R^2 = 0 oznacza, że model nie wyjaśnia żadnej zmienności w danych.
    R^2 = 1 oznacza, że model doskonale wyjaśnia zmienność w danych.

Wyniki walidacji krzyżowej (R^2) dla pięciu różnych iteracji. Oto interpretacja tych wyników:

    -1.10132456e+13: Model miał bardzo słabą wydajność w tej iteracji, a R^2 jest bardzo niskie, co sugeruje, że model źle dopasował się do danych.

    -6.73892362e+01: Model miał lepszą wydajność niż w pierwszej iteracji, ale R^2 nadal jest ujemne, co może oznaczać, że model nie pasuje dobrze do danych.

    4.97220934e-01: W tej iteracji model uzyskał pozytywną wartość R^2, co sugeruje, że model lepiej pasuje do danych niż w poprzednich iteracjach.

    -2.92894489e+13: Ta iteracja znowu wykazała bardzo niską wydajność modelu, a R^2 jest bardzo niskie.

    -2.25321442e+10: Wynik w tej iteracji również jest bardzo niski, co sugeruje, że model słabo pasuje do danych.

Średnia wartość R^2 z tych iteracji wynosi -8065045337269.175, co jest bardzo niską wartością. To sugeruje, że model regresji liniowej jest bardzo słaby w przewidywaniu cen nieruchomości na podstawie dostępnych atrybutów.

### Eksperyment w celu uzyskania lepszych danych

In [21]:
df.dropna(inplace=True)

# Wybierz tylko atrybuty numeryczne do normalizacji
numerical_features = ['floor', 'latitude', 'longitude', 'rooms', 'sq', 'year']

# Normalizacja atrybutów numerycznych
scaler = MinMaxScaler()
df[numerical_features] = scaler.fit_transform(df[numerical_features])

# Podziel dane na atrybuty i etykietę
X = df.drop(columns=['price'])
y = df['price']

# Podziel dane na zbiór treningowy i testowy
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Teraz możesz użyć X_train i y_train do trenowania modelu, a X_test i y_test do oceny jego wydajności.


In [23]:
model = LinearRegression()

# Przeprowadzenie 5-krotnej walidacji krzyżowej
scores = cross_val_score(model, X, y, cv=5, scoring='r2')

# Wyświetlenie wyników walidacji krzyżowej
print("Wyniki walidacji krzyżowej (R^2):", scores)
print("Średni wynik walidacji krzyżowej (R^2):", scores.mean())

Wyniki walidacji krzyżowej (R^2): [-6.84523456e+10 -6.73892362e+01  4.97220934e-01  2.08421703e-01
 -2.55731183e+13]
Średni wynik walidacji krzyżowej (R^2): -5128314123971.519


Obecne wyniki są znacznie lepsze, oznaczają lepszą zdolność modelu do dopasowania się do danych. Wartość średniego wyniku walidacji krzyżowej R^2 wynosi teraz -5128314123971.519, co wskazuje na to, że model lepiej wyjaśnia zmienność w danych w porównaniu do poprzednich wyników.

In [24]:
df = pd.read_csv("Houses.csv", encoding="windows-1250")

In [25]:
def preprocess_inputs(df):
    df = df.copy()
    
    # Drop unused columns
    df = df.drop(['Unnamed: 0', 'address', 'id'], axis=1)
    
    # Split df into X and y
    y = df['price']
    X = df.drop('price', axis=1)
    
    # Train-test split
    X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.7, shuffle=True, random_state=1)
    
    return X_train, X_test, y_train, y_test

In [27]:
X_train, X_test, y_train, y_test = preprocess_inputs(df)

In [28]:
X_train

city  floor   latitude  longitude  rooms     sq    year
23266    Kraków    3.0  50.046943  19.997153    1.0  23.75  2020.0
4050   Warszawa    0.0  52.322818  21.057657    4.0  76.60  2002.0
20631    Kraków    0.0  50.023016  19.908364    2.0  76.45  2018.0
23295  Warszawa    1.0  52.245964  21.133045    2.0  35.17  2022.0
23585  Warszawa    0.0  52.188721  21.058435    3.0  49.00  1970.0
...         ...    ...        ...        ...    ...    ...     ...
10955  Warszawa    4.0  52.277527  21.022353    2.0  37.50  1968.0
17289    Kraków    4.0  50.056192  19.928406    1.0  35.35  2020.0
5192   Warszawa    3.0  52.231958  21.006725    2.0  36.00  2020.0
12172    Poznań    0.0  52.387661  16.914801    1.0  20.00  1902.0
235    Warszawa    1.0  52.193987  21.045781    2.0  38.60  2020.0

[16634 rows x 7 columns]

In [29]:
y_train

23266    329000.0
4050     575000.0
20631    890000.0
23295    269051.0
23585    549000.0
           ...   
10955    468750.0
17289    618625.0
5192     459000.0
12172    187500.0
235      620000.0
Name: price, Length: 16634, dtype: float64

In [31]:
import numpy as np
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline

from sklearn.ensemble import GradientBoostingRegressor
nominal_transformer = Pipeline(steps=[
    ('onehot', OneHotEncoder(sparse=False))
])

preprocessor = ColumnTransformer(transformers=[
    ('nominal', nominal_transformer, ['city'])
], remainder='passthrough')

model = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('scaler', StandardScaler()),
    ('regressor', GradientBoostingRegressor())
])

In [32]:
model.fit(X_train, y_train)

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('nominal',
                                                  Pipeline(steps=[('onehot',
                                                                   OneHotEncoder(sparse=False))]),
                                                  ['city'])])),
                ('scaler', StandardScaler()),
                ('regressor', GradientBoostingRegressor())])

In [33]:
y_pred = model.predict(X_test)

rmse = np.sqrt(np.mean((y_test - y_pred)**2))
print("RMSE: {:.5f}".format(rmse))

baseline_errors = np.sum((y_test - np.mean(y_test))**2)
model_errors = np.sum((y_test - y_pred)**2)

r2 = 1 - (model_errors / baseline_errors)
print("R^2 Score: {:.5f}".format(r2))

RMSE: 245474.84277
R^2 Score: 0.78673


    RMSE (Root Mean Square Error):
        RMSE to miara, która określa, jak dokładnie model przewiduje wartości.
        RMSE oblicza pierwiastek średniego kwadratu różnicy między rzeczywistymi danymi (y_test) a przewidywanymi danymi (y_pred).
        Im niższa wartość RMSE, tym lepsza jakość modelu. W tym przypadku wynosi ona około 246098.14872, co oznacza, że model przewiduje średnio z błędem około 246098.15 jednostek w przypadku cen nieruchomości.

    R^2 Score (Coefficient of Determination):
        R^2 Score mierzy, jak dobrze model pasuje do danych w porównaniu do prostego modelu średniej (baseline).
        R^2 Score w zakresie od 0 do 1, gdzie 1 oznacza doskonałe dopasowanie modelu do danych, a 0 oznacza, że model nie jest lepszy od prostego modelu średniej.
        Wartość R^2 Score wynosi około 0.78564, co oznacza, że model wyjaśnia około 78.56% zmienności w danych cen nieruchomości.

Wynik R^2 Score na poziomie 0.78564 jest dość dobry, co sugeruje, że model regresji liniowej dobrze pasuje do danych. Jednak RMSE jest nadal dość wysoki, co oznacza, że istnieją pewne odchylenia między przewidywaniami modelu a rzeczywistymi danymi. 

In [34]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.metrics import mean_squared_error, r2_score

# Funkcja do przetwarzania danych
def preprocess_inputs(df):
    df = df.copy()
    
    # Drop unused columns
    df = df.drop(['Unnamed: 0', 'address', 'id'], axis=1)
    
    # Split df into X and y
    y = df['price']
    X = df.drop('price', axis=1)
    
    return X, y

# Wczytaj dane
df = pd.read_csv("Houses.csv", encoding="windows-1250")

# Usuń brakujące wartości
df.dropna(inplace=True)

# Podziel dane na atrybuty i etykietę
X, y = preprocess_inputs(df)

# Podziel dane na zbiór treningowy i testowy
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Preprocessing i model
nominal_transformer = Pipeline(steps=[('onehot', OneHotEncoder(sparse=False))])
preprocessor = ColumnTransformer(transformers=[('nominal', nominal_transformer, ['city'])], remainder='passthrough')
model = Pipeline(steps=[('preprocessor', preprocessor), ('scaler', StandardScaler()), ('regressor', GradientBoostingRegressor(n_estimators=200, max_depth=5, learning_rate=0.1, random_state=42))])

# Trenowanie modelu
model.fit(X_train, y_train)

# Przewidywanie
y_pred = model.predict(X_test)

# Ocena modelu
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
r2 = r2_score(y_test, y_pred)

print("RMSE: {:.5f}".format(rmse))
print("R^2 Score: {:.5f}".format(r2))


RMSE: 209474.11972
R^2 Score: 0.83332


Wprowadziliśmy kilka zmian w kodzie, takie jak dostosowanie podziału danych na treningowe i testowe (30% testowych) oraz dostosowanie parametrów modelu GradientBoostingRegressor, takich jak n_estimators, max_depth, learning_rate. Te parametry mogą wymagać dalszego tuningu, ale te wartości wydają się sensowne na początek.Otrzymane RMSE na poziomie 209474.11972 i R^2 Score wynoszący 0.83332 sugerują, że model jest teraz bardziej dokładny w przewidywaniu cen nieruchomości.